File: latihan/2-inheritance/KoleksiKendaraan.cpp
/*  NIM     : 13521129
    Nama    : Chiquita Ahsanunnisa
    Tanggal : 19 Februari 2023
    Topik   : Latihan Inheritance - KoleksiKendaraan.cpp
*/

#include "KoleksiKendaraan.hpp"

// KELAS KOLEKSI KENDARAAN

// CTOR
KoleksiKendaraan::KoleksiKendaraan() : size(100), nEff(0) {
    this->buffer = new Kendaraan*[this->size];
}
KoleksiKendaraan::KoleksiKendaraan(int sz) : size(sz), nEff(0) {
    this->buffer = new Kendaraan*[this->size];
}

// CCTOR
KoleksiKendaraan::KoleksiKendaraan(const KoleksiKendaraan& kk) {
    this->size = kk.size;
    this->nEff = kk.nEff;
    this->buffer = new Kendaraan*[this->size];
    for (int i = 0; i < this->nEff; i++) {
        this->buffer[i] = kk.buffer[i]->clone();
    }
}

// ASSIGNMENT
KoleksiKendaraan& KoleksiKendaraan::operator=(const KoleksiKendaraan& kk) {
    for (int i = 0; i < this->nEff; i++) {
        delete this->buffer[i];
    }
    delete [] this->buffer;

    this->size = kk.size;
    this->nEff = kk.nEff;
    this->buffer = new Kendaraan*[this->size];
    for (int i = 0; i < this->nEff; i++) {
        this->buffer[i] = kk.buffer[i]->clone();
    }
    return *this;
}

// DTOR
KoleksiKendaraan::~KoleksiKendaraan() {
    for (int i = 0; i < this->nEff; i++) {
        delete this->buffer[i];
    }
    delete [] this->buffer;
}

// OTHERS
void KoleksiKendaraan::printAll() const {
    for (int i = 0; i < this->nEff; i++) {
        cout << "------------------------" << endl;
        this->buffer[i]->printInfo();
    }
}
KoleksiKendaraan& KoleksiKendaraan::operator<<(const Kendaraan& k) {
    this->buffer[this->nEff] = k.clone();
    this->nEff++;
    return *this;
}
KoleksiKendaraan& KoleksiKendaraan::operator<<(const KoleksiKendaraan& kk) {
    int batas = (this->size > this->nEff + kk.nEff ? kk.nEff : this->size - this->nEff);
    for (int i = 0; i < batas; i++) {
        this->buffer[this->nEff + i] = kk.buffer[i]->clone();
    }
    this->nEff += batas;
    return *this;
}

File: latihan/5-java-inheritance/KoleksiKendaraan.java
/*  NIM     : 13521129
    Nama    : Chiquita Ahsanunnisa
    Tanggal : 31 Maret 2023
    Topik   : Latihan Java Inheritance - KoleksiKendaraan.java
*/

public class KoleksiKendaraan {
    private Kendaraan[] buffer;
    private int nEff;

    // CTOR
    public KoleksiKendaraan() {
        buffer = new Kendaraan[100];
        nEff = 0;
    }
    public KoleksiKendaraan(int size) {
        buffer = new Kendaraan[size];
        nEff = 0;
    }
    public KoleksiKendaraan(KoleksiKendaraan other) {
        this.nEff = other.nEff;
        this.buffer = new Kendaraan[other.buffer.length];
        for (int i = 0; i < this.nEff; i++) {
            this.buffer[i] = other.buffer[i].copy();
        }
    }

    // OTHERS
    public void printAll() {
        for (int i = 0; i < nEff; i++) {
            buffer[i].printInfo();
            System.out.println("-------------------------------");
        }
    }
    public void insert(Kendaraan kendaraan) {
        buffer[nEff] = kendaraan.copy();
        nEff++;
    }
    public void insert(KoleksiKendaraan other){
        int batas = (this.buffer.length > nEff + other.nEff ? other.nEff : this.buffer.length - this.nEff);
        for (int i = 0; i < batas; i++) {
            this.buffer[this.nEff + i] = other.buffer[i].copy();
        }
        this.nEff += batas;
    }
}
